### PR TITLE
Apply platforms to resolved classpaths instead of dependency containe…

### PR DESCRIPTION
**Pull request by anuraaga**
…rs and have maven publication use resolution result for all variants.

This is an alternative to #1281. I think it's better since BOMs aren't in the published POMs at all - the platforms are used to manage versions for this project but aren't meant for users. It removes worries related to gRPC version, etc too for free.

API POM - no BOMs
```
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>io.opentelemetry</groupId>
  <artifactId>opentelemetry-api</artifactId>
  <version>0.5.0-SNAPSHOT</version>
  <name>OpenTelemetry Java</name>
  <description>OpenTelemetry API</description>
  <url>https://github.com/open-telemetry/opentelemetry-java</url>
  <licenses>
    <license>
      <name>The Apache License, Version 2.0</name>
      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
    </license>
  </licenses>
  <developers>
    <developer>
      <id>opentelemetry</id>
      <name>OpenTelemetry Gitter</name>
      <url>https://gitter.im/open-telemetry/community</url>
    </developer>
  </developers>
  <scm>
    <connection>scm:git:git@github.com:open-telemetry/opentelemetry-java.git</connection>
    <developerConnection>scm:git:git@github.com:open-telemetry/opentelemetry-java.git</developerConnection>
    <url>git@github.com:open-telemetry/opentelemetry-java.git</url>
  </scm>
  <dependencies>
    <dependency>
      <groupId>javax.annotation</groupId>
      <artifactId>javax.annotation-api</artifactId>
      <version>1.3.2</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>io.opentelemetry</groupId>
      <artifactId>opentelemetry-context-prop</artifactId>
      <version>0.5.0-SNAPSHOT</version>
      <scope>compile</scope>
    </dependency>
  </dependencies>
</project>
```

SDK POM - Guava version filled in
```
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>io.opentelemetry</groupId>
  <artifactId>opentelemetry-sdk</artifactId>
  <version>0.5.0-SNAPSHOT</version>
  <name>OpenTelemetry Java</name>
  <description>OpenTelemetry SDK</description>
  <url>https://github.com/open-telemetry/opentelemetry-java</url>
  <licenses>
    <license>
      <name>The Apache License, Version 2.0</name>
      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
    </license>
  </licenses>
  <developers>
    <developer>
      <id>opentelemetry</id>
      <name>OpenTelemetry Gitter</name>
      <url>https://gitter.im/open-telemetry/community</url>
    </developer>
  </developers>
  <scm>
    <connection>scm:git:git@github.com:open-telemetry/opentelemetry-java.git</connection>
    <developerConnection>scm:git:git@github.com:open-telemetry/opentelemetry-java.git</developerConnection>
    <url>git@github.com:open-telemetry/opentelemetry-java.git</url>
  </scm>
  <dependencies>
    <dependency>
      <groupId>javax.annotation</groupId>
      <artifactId>javax.annotation-api</artifactId>
      <version>1.3.2</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>io.opentelemetry</groupId>
      <artifactId>opentelemetry-api</artifactId>
      <version>0.5.0-SNAPSHOT</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>com.google.guava</groupId>
      <artifactId>guava</artifactId>
      <version>28.2-android</version>
      <scope>runtime</scope>
    </dependency>
  </dependencies>
</project>
```

### CODE REVIEW

**Comment by iNikem** [comment]
https://oss.jfrog.org/artifactory/oss-snapshot-local/io/opentelemetry/opentelemetry-api/0.5.0-SNAPSHOT/

module metadata does exist in this repo.

File: build.gradle [hunk 1]
```diff
@@ -179,11 +179,14 @@ configure(opentelemetryProjects) {
     }
 
     dependencies {
-
-        implementation platform(boms.grpc),
-                platform(boms.guava),
-                platform(boms.protobuf),
-                platform(boms.zipkin_reporter)
+        configurations.all {
+            if (it.name.endsWith('Classpath')) {
```
Comment (iNikem) [inline build.gradle:183]: One usually does not add anything directly to `XXClasspath` configuration. Why is this needed in this case?

Comment (anuraaga) [reply build.gradle:183]: The classpath configurations are only for resolving dependencies, not publishing them. Since we want to use the BOMs for resolving, not publishing (as we see in the large thread on the other PR ;) ), it makes sense to add them to the classpath configurations instead.

This is the recommended approach by Gradle for this use case - https://github.com/gradle/gradle/issues/10861#issuecomment-576562961

Comment (iNikem) [reply build.gradle:183]: Isn't it enough to use `fromResolutionResult` as you did below?

Comment (anuraaga) [reply build.gradle:183]: Nah, at first I thought so too but while working on this realized it's a different topic.

- `fromResolutionResult` fills in version numbers for dependencies resolved via bom. So for `guava` dependency, without it, it would be present but have no version number

- Moving platform from "dependency bag" configurations to classpath configurations. Ensures BOM imports are not added to the published POM which is a good thing to reduce the complexity of the POM, we want the BOMs to resolve the version numbers, not to publish them

File: build.gradle [hunk 2]
```diff
@@ -179,11 +179,14 @@ configure(opentelemetryProjects) {
...
+                add(it.name, platform(boms.grpc))
+                add(it.name, platform(boms.guava))
+                add(it.name, platform(boms.protobuf))
+                add(it.name, platform(boms.zipkin_reporter))
```
Comment (DotSpy) [inline build.gradle:187]: maybe we should use here forcedPlatform instead of platform what's ur opinion @anuraaga

Comment (anuraaga) [reply build.gradle]: Yeah I usually use enforcedPlatform for this sort of management - switched to it

**Comment by DotSpy** [comment]
LGTM

**Comment by anuraaga** [comment]
In #1289 noticed we have to add dependencies both on artifacts and platform in subprojects now, for example for Guava. This seems both tedious and against the spirit of dependency management for a Java project, nor does it have to be that way...

Hope people can look at this change again with an open mind. Thanks.

**Comment by jkwatson** [comment]
This needs a rebase, now that #1281 has been merged

**Comment by anuraaga** [comment]
Thanks! Rebased

**Comment by jkwatson** [comment]
@iNikem and @DotSpy Can you take another look at this and approve if you think it is good to go? Thanks!

**Comment by DotSpy** [comment]
LGTM

**Comment by anuraaga** [comment]
@bogdanrutu perhaps merge this before the release to get the new POMs?

**Comment by anuraaga** [comment]
@bogdandrutu Sorry I noticed I missed a character in the `@mention` (was on mobile so no autocomplete ><). Guess this missed the release but can we merge now?

**Comment by anuraaga** [comment]
@jkwatson both approved so think we can merge this?

**Comment by jkwatson** [comment]
> @jkwatson both approved so think we can merge this?

It looks like one unresolved conversation is still outstanding. Can y'all resolve it if it's all good? Thanks!
